lib/worker.js

```diff
@@ -0,0 +1,211 @@
+  this._checkAlive();
+  this._workerContext.ref();
+};
+
+Worker.prototype.unref = function() {
+  this._checkAlive();
+  this._workerContext.unref();
+};
+
+if (process.isWorkerInstance) {
+  const postMessage = function(payload, transferList, type) {
+    if (transferList == null)
+      transferList = EMPTY_ARRAY;
+    else if (!Array.isArray(transferList))
+      throw new Error('transferList must be an array');
+
+    WorkerContextBinding.workerWrapper._postMessage(JSONStringify(payload),
+                                                    transferList,
+                                                    type);
+  };
+  const workerFatalError = function(er) {
+    var defaultStack = null;
+    var defaultMessage = '[to string conversion failed]';
+    var defaultBuiltinType = 'Error';
+
+    var message = defaultMessage;
+    var builtinType = defaultBuiltinType;
+    var stack = defaultStack;
+    var additionalProperties = {};
+
+    if (er instanceof Error) {
+      try {
+        builtinType = er.name;
+        if (typeof builtinType !== 'string')
+          builtinType = defaultBuiltinType;
+      } catch (ignore) {}
+
+      try {
+        stack = er.stack;
+        if (typeof stack !== 'string')
+          stack = defaultStack;
+      } catch (ignore) {}
+      try {
+        // Get inherited enumerable properties.
+        // .name, .stack and .message are all non-enumerable
+        for (var key in er)
+          additionalProperties[key] = er[key];
+        // The main message delivery must always succeed.
+        JSONStringify(additionalProperties);
+      } catch (e) {
+        additionalProperties = {};
+      }
+    }
+
+    try {
+      if (er instanceof Error) {
+        message = er.message;
+        if (typeof message !== 'string')
+          message = '' + er;
+      } else {
+        message = '' + er;
+      }
+    } catch (e) {
+      message = defaultMessage;
+    }
+
+    postMessage({
+      message: message,
+      stack: stack,
+      additionalProperties: additionalProperties,
+      builtinType: builtinType
+    }, null, WorkerContextBinding.EXCEPTION);
+  };
+
+  util._extend(Worker, EventEmitter.prototype);
+  EventEmitter.call(Worker);
+
+  WorkerContextBinding.workerWrapper._onmessage =
+      function(payload, messageType) {
+    payload = JSONParse(payload);
+    switch (messageType) {
+      case WorkerContextBinding.USER:
+        return Worker.emit('message', payload);
+      case WorkerContextBinding.INTERNAL:
+        assert.fail('unreachable');
+      case WorkerContextBinding.EXCEPTION:
+        assert.fail('unreachable');
+      default:
+        assert.fail('unreachable');
+    }
+  };
+
+  Worker.postMessage = function(payload) {
+    postMessage(payload, null, WorkerContextBinding.USER);
+  };
+
+  Object.defineProperty(Worker, '_workerFatalError', {
+    configurable: true,
+    writable: false,
+    enumerable: false,
+    value: workerFatalError
+  });
+}
+
+
+module.exports = Worker;
```

Comment:
On a general note, maybe it's a good idea to use ES6 symbols for private properties here, to avoid them leaking into user land.